### PR TITLE
Time-dependent path graphics

### LINKED ISSUE #13320: Time-dependent (interval) PathGraphics
Enable different segments of the same orbit path, representing different segments of time, to have different graphical properties. Currently, you can apply different properties with a `TimeIntervalCollection`, but they apply to the entire path during that interval.

### PR DESCRIPTION
**Pull request by eringram**
# Description

image

This PR adds a new concept of a material mode that affects how path graphics of an entity are rendered. This is done by adding a new property called `materialMode` to the `PathGraphics` class which can have the values "WHOLE" or "PORTIONS".

The "WHOLE" mode works the same as the existing behavior: when an entity/CZML object's path material changes over time via an interval or if it is a sampled property, the entire path's material changes as well.

The new "PORTIONS" mode: when an entity's path material changes over time, the path is instead split into segments based on the different intervals. This allows you to see different material properties from past and future timestamps.

In the following screenshots, the left column is "WHOLE" mode/previous behavior, the right column is "PORTIONS":

image

## Issue number and link

Fixes https://github.com/CesiumGS/cesium/issues/13320

## Testing plan

- Added a sandcastle testing several scenarios, including edge cases like using PORTIONS mode with non-time-dynamic materials
- Added unit tests to PathVisualizerSpec.js for handling the new property
- TODO: I am working on creating an example with many intervals that will cause the new mode to create many segements/a very large `PolylineCollection` to test the performance.

## Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

## AI acknowledgment

- [x] I used AI to generate content in this PR
- [x] If yes, I have reviewed the AI-generated content before submitting

If yes, I used the following Tools(s) and/or Service(s):

GitHub Copilot

If yes, I used the following Model(s):

Claude Opus 4.8, GPT-5.3-Codex

### CODE REVIEW

**Comment by eringram** [comment]
@lukemckinstry this is (finally) ready for review. Don't worry about the huge amount of lines added, most of it is the CZML sample data. The most important parts are in Source/DataSources/, particularly PathVisualizer.js

File: packages/engine/Source/DataSources/PathVisualizer.js [hunk 1]
```diff
@@ -888,11 +996,190 @@ PolylineUpdater.prototype.updateObject = function (time, item) {
...
+    }
+
+    // Prevent non-positive split steps from creating non-terminating loops.
+    // Positive fractional resolutions are valid; only values <= 0 fall back to the default.
+    const splitResolution = resolution > 0 ? resolution : defaultResolution;
+
+    const intervals = materialProp.intervals;
+    let nextSegIndex = 0;
+
+    if (!defined(intervals)) {
```
Comment (lukemckinstry) [inline packages/engine/Source/DataSources/PathVisualizer.js:1030]: Do we have any sample data for this code path? It seems like all 4 czml datasets have multiple intervals in path.material. I do see test coverage for this in the specs, but perhaps for learning and docs it would be good to add sample data for this and include it in the sandcastle.

File: packages/engine/Source/DataSources/PathVisualizer.js [hunk 2]
```diff
@@ -888,11 +996,190 @@ PolylineUpdater.prototype.updateObject = function (time, item) {
...
   polyline.positions = positions;
+
   polyline.material = MaterialProperty.getValue(
     time,
     pathGraphics._material,
     polyline.material,
   );
+
+  const materialMode = Property.getValueOrUndefined(
+    pathGraphics._materialMode,
```
Comment (lukemckinstry) [inline packages/engine/Source/DataSources/PathVisualizer.js]: I think it makes sense to use the public API properties for _material and _materialMode  here and throughout this class

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js [hunk 3]
```diff
@@ -1267,6 +1286,726 @@ describe(
...
+      path.trailTime = new ConstantProperty(10);
+      path.material = new ColorMaterialProperty(sampledColor);
+
+      expect(function () {
+        visualizer.update(updateTime);
+      }).not.toThrow();
+
+      const shownPolylines = getShownPolylines(scene);
+      expect(shownPolylines.length).toBeGreaterThan(1);
+      expect(shownPolylines[0].material.uniforms.color).toBeDefined();
```
Comment (lukemckinstry) [inline packages/engine/Specs/DataSources/PathVisualizerSpec.js]: Could and should this be checking the different sampled colors in addition to there being multiple polylines? 

EDIT: I see some later tests cover this, up to you if it would help here.

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js [hunk 4]
```diff
@@ -1267,6 +1286,726 @@ describe(
...
+      path.width = new ConstantProperty(8.0);
+      path.resolution = new ConstantProperty(5.0);
+      path.leadTime = new ConstantProperty(10);
+      path.trailTime = new ConstantProperty(10);
+      path.material = material;
+
+      visualizer.update(updateTime);
+
+      const shownPolylines = getShownPolylines(scene);
+      expect(shownPolylines.length).toBeGreaterThan(1);
```
Comment (lukemckinstry) [inline packages/engine/Specs/DataSources/PathVisualizerSpec.js:1508]: Same question as above, could and should this measure for something like color in addition to there being multiple polylines?

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js [hunk 5]
```diff
@@ -1267,6 +1286,726 @@ describe(
...
+      path.material = new ColorMaterialProperty(sampledColor);
+
+      expect(function () {
+        visualizer.update(updateTime);
+      }).not.toThrow();
+
+      expect(getShownPolylines(scene).length).toBeGreaterThan(0);
+    });
+
+    it("clips interval colors to the active lead/trail window", function () {
```
Comment (lukemckinstry) [inline packages/engine/Specs/DataSources/PathVisualizerSpec.js]: I'm not sure the title and content of this test match. "clips" seems to indicate we want to test a limited window to verify we are not showing segments we do not want shown. But the sample time is such that both interval segments are visible and the test checks that both are visible. Should we test at a time when one is supposed to be not visible?

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js [hunk 6]
```diff
@@ -1267,6 +1286,726 @@ describe(
...
+          iso8601: "2026-04-01T00:00:00Z/2026-04-01T00:00:20Z",
+          data: new ColorMaterialProperty(Color.RED),
+        }),
+      );
+      path.material = material;
+
+      visualizer.update(JulianDate.fromIso8601("2026-04-01T00:00:08Z"));
+      expect(getShownPolylines(scene).length).toBeGreaterThan(0);
+
+      visualizer.update(JulianDate.fromIso8601("2026-04-01T00:00:15Z"));
```
Comment (lukemckinstry) [inline packages/engine/Specs/DataSources/PathVisualizerSpec.js:1881]: So the line segment gets shown even if the entity is not available at that time? That seems to be what  the spec is calling for, but I wanted to clarify that is intended design,

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js [hunk 7]
```diff
@@ -1267,6 +1286,726 @@ describe(
...
+        }),
+      );
+      path.material = material;
+
+      visualizer.update(JulianDate.fromIso8601("2026-04-01T00:00:05Z"));
+      expect(getShownPolylines(scene).length).toBeGreaterThan(1);
+
+      visualizer.update(JulianDate.fromIso8601("2026-04-01T00:00:15Z"));
+      const shownConstantPolylines = getShownPolylines(scene);
+      expect(shownConstantPolylines.length).toBeGreaterThan(0);
```
Comment (lukemckinstry) [inline packages/engine/Specs/DataSources/PathVisualizerSpec.js]: why greaterThan(0) and not equals(1) here?

**Review by lukemckinstry — COMMENTED** [review]
Looks great! Some small comments - mostly questions and suggestions.

File: packages/engine/Source/DataSources/PathVisualizer.js (hunk 2 above)
Comment (eringram) [reply packages/engine/Source/DataSources/PathVisualizer.js]: I agree, updated to use the public API for these

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js (hunk 3 above)
Comment (eringram) [reply packages/engine/Specs/DataSources/PathVisualizerSpec.js]: Yes, I added a check just to make sure the colors are distinct

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js (hunk 4 above)
Comment (eringram) [reply packages/engine/Specs/DataSources/PathVisualizerSpec.js:1508]: Same as above

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js (hunk 5 above)
Comment (eringram) [reply packages/engine/Specs/DataSources/PathVisualizerSpec.js]: Good point, updated this to test at a better timestamp where only one is visible

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js (hunk 6 above)
Comment (eringram) [reply packages/engine/Specs/DataSources/PathVisualizerSpec.js:1881]: No, if the entity is unavailable the segment shouldn't be shown: there was a bug where the paths of previously available, now unavailable entities were not cleaned up. Good catch, thanks-- just pushed a fix for it

File: packages/engine/Specs/DataSources/PathVisualizerSpec.js (hunk 7 above)
Comment (eringram) [reply packages/engine/Specs/DataSources/PathVisualizerSpec.js]: This comment made me realize that this test isn't very useful, because the lead and trail times are long enough that all three segments are visible for the entire time period, and the object's availability doesn't change like in the test right above this one. I think each individual piece of the behavior is tested elsewhere, so I just removed it to keep this file simpler

File: packages/engine/Source/DataSources/PathVisualizer.js (hunk 1 above)
Comment (eringram) [reply packages/engine/Source/DataSources/PathVisualizer.js:1030]: The "sampled flight path" data doesn't have explicitly defined intervals, but uses a [sampled property](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Structure#sampled-property-values) with an epoch (so intervals are interpolated/created at runtime basically). So `!defined(intervals)` would be true for that case and it goes down this path.

For a constant property that doesn't have intervals and isn't sampled, it skips this entire block earlier due to this condition on line 1017 because `materialProp.isConstant = true`:
```
if (materialMode === PathMode.PORTIONS && !materialProp.isConstant) {...
```

To provide a constant property example, I added another dataset with a constant material which shows its path is unaffected when you switch between WHOLE and PORTIONS mode

Comment (lukemckinstry) [reply packages/engine/Source/DataSources/PathVisualizer.js:1030]: I am curious what is the expected behavior when using this code path- specifically sampled property without intervals defined in the czml.

Can one create a multi-colored time dependent path using a sampled property without defining intervals in the czml? If yes I think that would be good to demonstrate for users in the sandcastle.

Comment (lukemckinstry) [reply packages/engine/Source/DataSources/PathVisualizer.js:1030]: Discussed offline with @eringram, this use case is covered by sampled CZML example in sandcastle!

**Review by lukemckinstry — APPROVED** [review]
Looks great 🚀
